src/lib/api/event.recorder.ts

```diff
@@ -189,48 +185,52 @@ export function handleEvents(prebootData: PrebootData,
     return;
   }
 
-  // get all nodes under the server root that match the given selector
-  const nodes: NodeListOf<Element> = serverRoot.querySelectorAll(eventSelector.selector);
-
-  // don't do anything if no nodes found
-  if (!nodes) {
-    return;
-  }
-
-  // we want to add an event listener for each node and each event
-  for (let i = 0; i < nodes.length; i++) {
-    const node = nodes.item(i);
-    eventSelector.events.forEach((eventName: string) => {
-      // get the appropriate handler and add it as an event listener
-      const handler = createListenHandler(prebootData, eventSelector, appData, node);
-      node.addEventListener(eventName, handler);
-
-      // need to keep track of listeners so we can do node.removeEventListener()
-      // when preboot done
-      if (prebootData.listeners) {
-        prebootData.listeners.push({
-          node: node as HTMLElement,
-          eventName,
-          handler
-        });
-      }
-    });
-  }
+  // Attach delegated event listeners for each event selector.
+  // We need to use delegated events as only the top level server node
+  // exists at this point.
+  eventSelector.events.forEach((eventName: string) => {
+    // get the appropriate handler and add it as an event listener
+    const handler = createListenHandler(_document, prebootData, eventSelector, appData);
+    _document.addEventListener(eventName, handler);
```

Comment:
The second one if it works properly would be the way to go

Reply:
One other thing - we should probably use the capture phase instead of the bubble one. Otherwise if any of the respective event handlers call `stopPropagation()` it will not get to the server node & the event won't be captured.